src/TimeModeling/Modeling/distributed.jl

```diff
@@ -0,0 +1,243 @@
+end
+
+
+"""
+    single_reduce!(x, y) 
+
+Inplace reduction of y into x.
+"""
+@inline single_reduce!(x, y) = begin x .+= y; nothing end
+
+for T in [judiVector, judiWeights, judiWavefield]
+    @eval @inline single_reduce!(x::$T, y) = begin push!(x, y); nothing end
+end
+
+@inline single_reduce!(x::T, y) where {T<:Number} = begin x += y; nothing end
+
+@inline function single_reduce!(x::Tuple, y::Tuple)
+    nx = length(x)
+    ny = length(y)
+    (nx == ny) || throw(ArgumentError("Incompatible lengths ($nx, $ny)"))
+    for (xi, yi) in zip(x, y)
+        single_reduce!(xi, yi)
+    end
+    nothing
+end
+
+"""
+    safe_gc()
+
+Generic GC, compatible with different julia versions of it.
+"""
+safe_gc() = try Base.GC.gc(); catch; gc() end
+
+"""
+    local_reduce!(future, other)
+
+Reduce `other` future into local `future`. This is perform remotely on the julia worker
+`future.where`.
+"""
+function local_reduce!(my_future::_TFuture, other_future::_TFuture)
+    y = remotecall_fetch(fetch, other_future.where, other_future)
+    x = fetch(my_future)
+    single_reduce!(x, y)
+    # Force GC since julia parallel still has issues with it
+    y=1;safe_gc(); remotecall(safe_gc, other_future.where);
+    nothing
+end
+
+"""
+    reduce_level!(futures, nleaf)
+
+Reduce one level of the reduction tree consisting of `nleaf` futures. Each `leaf`
+reduces into itself `futures[i]`.
+"""
+function reduce_level!(futures::Vector{_TFuture}, nleaf::Int)
+    nleaf == 0 && return
+    @sync for i = 1:nleaf
+        @async remotecall_fetch(local_reduce!, futures[i].where, futures[i], futures[nleaf+i])
+    end
+end
+
+"""
+    reduce!(x)
+binary-tree reduction of a Vector of Futures `x`
+
+Modified from `DistributedOperations.jl` (Striped from custom types in it) and extended to multiple outputs
+with different reduction functions.
+
+Copyright (c) 2020: Chevron U.S.A Inc.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+function reduce!(futures::Vector{_TFuture})
```

Comment:
Just out of curiosity: why is implementation of reduce! here needed? Is there anything wrong with Julia's default reduce?

Reply:
THere is no mapreduce in julia, this does the reduction as efficiently as possible (only local reduction on workers, the final small reduction on main worker rather than gather everything then reduce. Lot less comm